README.md
# GOVAPID

[![GoDoc][godoc-image]][godoc-url]
[![codecov](https://codecov.io/gh/abdullahdiaa/govapid/branch/main/graph/badge.svg?token=XY7LZ4XCK3)](https://codecov.io/gh/abdullahdiaa/govapid)
[![Build Status](https://www.travis-ci.com/AbdullahDiaa/GoVAPID.svg?branch=main)](https://www.travis-ci.com/AbdullahDiaa/GoVAPID)
[![Go Report Card](https://goreportcard.com/badge/github.com/AbdullahDiaa/govapid)](https://goreportcard.com/report/github.com/AbdullahDiaa/govapid)

> A micro-package to generate VAPID public and private keys and VAPID authorization headers, required for sending web push notifications.
> The library only supports VAPID-draft-02+ specification.

## Usage

```go
package main

import (
	"fmt"
	"github.com/AbdullahDiaa/govapid"
)

func main() {
	//Generate VAPID keys
	VAPIDkeys, err := govapid.GenerateVAPID()
	if err != nil {
		fmt.Println(err)
	}
	fmt.Printf("Public Key:%s\nPrivate Key:%s", VAPIDkeys.Public, VAPIDkeys.Private)

	//Generate VAPID Authorization header which contains JWT signed token and VAPID public key
	subURL, _ := url.Parse("https://fcm.googleapis.com/fcm/send/d5E6exZV5dM:APA91bHI09qFrkxTShu_pUVk-7ZukjdVhEJeZNUt29hSeBez93KlgXDK6Y9BThZMfWUqGhQ8yiWzYqT1gIGUxA5DVuwuARpJPSzk5XFp3yR1kepLKWOOdIgcAO6GRGoZYngmAFc6oufU")

	claims := map[string]interface{}{
		"aud": fmt.Sprintf("%s://%s", subURL.Scheme, subURL.Host),
		"exp": time.Now().Add(time.Hour * 12).Unix(),
		"sub": "mailto:mail@mail.com"}

	AuthorizationHeader, _ := GenerateVAPIDAuth(VAPIDkeys, claims)
	fmt.Println(AuthorizationHeader)
}

```

## Documentation

You can view detailed documentation here: [GoDoc][godoc-url].

## Contributing

There are many ways to contribute:
- Fix and [report bugs](https://github.com/AbdullahDiaa/GoVAPID/issues/new)
- [Improve documentation](https://github.com/AbdullahDiaa/GoVAPID/issues?q=is%3Aopen+label%3Adocumentation)
- [Review code and feature proposals](https://github.com/AbdullahDiaa/GoVAPID/pulls)


## Changelog

View the [changelog](/CHANGELOG.md) for the latest updates and changes by
version.

## License

[Apache License 2.0][licence-url]

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

[godoc-image]: https://godoc.org/github.com/AbdullahDiaa/GoVAPID?status.svg
[godoc-url]: https://godoc.org/github.com/AbdullahDiaa/GoVAPID
[licence-url]: https://github.com/AbdullahDiaa/GoVAPID/blob/main/LICENSE

CHANGELOG.md
## v1.0.1-rc1 [2021-06-15]

#### Release Notes
- Add support for generating VAPID authorization header

govapid_test.go
package govapid

import (
	"encoding/base64"
	"fmt"
	"testing"
	"time"
)

const succeed = "\u2705"
const failed = "\u274C"

//TestGenerateVAPID will test VAPID Keys generation
func TestGenerateVAPID(t *testing.T) {
	t.Log("Make sure correct Public and Private VAPID keys are generated")
	{
		VAPIDKeys, err := GenerateVAPID()
		if err != nil {
			t.Fatalf("\t%s\tShould generate VAPID keys, got an error %s", failed, err.Error())
		}
		t.Logf("\t%s\tPublic key and Private key generated successfully", succeed)

		//Validate keys length
		if len(VAPIDKeys.Public) != 87 {
			t.Fatalf("\t%s\tInvalid Public VAPID key", failed)
		}
		t.Logf("\t%s\tPrivate key length is valid", succeed)

		if len(VAPIDKeys.Private) != 43 {
			t.Fatalf("\t%s\tInvalid Private VAPID key", failed)
		}
		t.Logf("\t%s\tPublic key length is valid", succeed)

		//Test base64 decoding
		_, err = base64.RawURLEncoding.DecodeString(VAPIDKeys.Private)
		if err != nil {
			t.Fatalf("\t%s\tInvalid Private key: %s", failed, err.Error())
		}
		t.Logf("\t%s\tPrivate key is valid", succeed)

		_, err = base64.RawURLEncoding.DecodeString(VAPIDKeys.Public)
		if err != nil {
			t.Fatalf("\t%s\tInvalid Public key: %s", failed, err.Error())
		}
		t.Logf("\t%s\tPublic key is valid", succeed)
	}
}

func TestVerifyClaims(t *testing.T) {

	t.Log("Try with valid claims map, Should pass")
	{
		claims := map[string]interface{}{
			"aud": "https://fcm.googleapis.com",
			"exp": time.Now().Add(time.Hour * 12).Unix(),
			"sub": fmt.Sprintf("mailto:mail@mail.com")}

		err := verifyClaims(claims)
		if err != nil {
			t.Fatalf("\t%s\tShould be valid claim got an error instead: %s", failed, err.Error())
		}
	}

	t.Logf("\t%s\tA claim with all fields valid", succeed)

	t.Log("Expiry date should be within 24 hours")
	{
		claims := map[string]interface{}{
			"aud": "https://fcm.googleapis.com",
			"exp": time.Now().Add(time.Hour * -12).Unix(),
			"sub": fmt.Sprintf("mailto:mail@mail.com")}

		err := verifyClaims(claims)
		if err.Error() != "expiry claim (exp) already expired" {
			t.Fatalf("\t%s\tShould return an error with the expired date got:%s", failed, err.Error())
		}
		t.Logf("\t%s\tInvalid exp claim (early)", succeed)

		//More than 24 hours
		claims["exp"] = time.Now().Add(time.Hour * 25).Unix()

		err = verifyClaims(claims)
		if err.Error() != "expiry claim (exp) maximum value is 24 hours" {
			t.Fatalf("\t%s\tShould return an error with max value of 24 hours got:%s", failed, err.Error())
		}
		t.Logf("\t%s\tInvalid exp claim (late)", succeed)
	}

	t.Log("Subscriber claim should be an email starting with (mailto) or URL")
	{
		claims := map[string]interface{}{
			"aud": "https://fcm.googleapis.com",
			"exp": time.Now().Add(time.Hour * -12).Unix(),
			"sub": "mail@mail.com"}

		err := verifyClaims(claims)
		if err.Error() != "“Subscriber” claim (sub) is invalid, it should be an email or contact URL" {
			t.Fatalf("\t%s\tShould return an error with invalid sub got:%s", failed, err.Error())
		}
		t.Logf("\t%s\tinvalid email format for Subscriber” field", succeed)

		//non-https URL
		claims["sub"] = "http://push-service.com"

		err = verifyClaims(claims)
		if err.Error() != "“Subscriber” claim (sub) is invalid, it should be an email or contact URL" {
			t.Fatalf("\t%s\tShould return an error with invalid sub got:%s", failed, err.Error())
		}
		t.Logf("\t%s\tinvalid URL (non-https) format for Subscriber” field", succeed)
	}

}

func TestGenerateVAPIDAuth(t *testing.T) {

}

func ExampleGenerateVAPID() {
	VAPIDKeys, err := GenerateVAPID()
	if err != nil {
		fmt.Println(err)
	}

	fmt.Println(len(VAPIDKeys.Public), len(VAPIDKeys.Private))
	// Output:
	// 87 43
}


govapid.go
//Package govapid is a micro-package to generate VAPID public and private keys and VAPID authentication headers, required for sending web push notifications.
//The library only supports VAPID-draft-02+ specification.
// https://datatracker.ietf.org/doc/html/draft-ietf-webpush-vapid-02
package govapid

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	_ "crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"strings"
	"time"
)

//VAPIDKeys contains the public and private VAPID keys
type VAPIDKeys struct {
	Public  string
	Private string
}

//GenerateVAPID will generate public and private VAPID keys
func GenerateVAPID() (VAPIDKeys, error) {

	curve := elliptic.P256()
	privateKey, x, y, err := elliptic.GenerateKey(curve, rand.Reader)
	if err != nil {
		return VAPIDKeys{}, err
	}

	publicKey := elliptic.Marshal(curve, x, y)

	privKey := base64.RawURLEncoding.EncodeToString(privateKey)
	pubKey := base64.RawURLEncoding.EncodeToString(publicKey)

	return VAPIDKeys{pubKey, privKey}, nil
}

// validateVAPIDKeys will validate the length and encoding of VAPID keys
func validateVAPIDKeys(keys VAPIDKeys) error {
	if len(keys.Public) != 87 {
		return errors.New("invalid Public key length")
	}

	if len(keys.Private) != 43 {
		return errors.New("invalid Private key length")
	}

	_, err := base64.RawURLEncoding.DecodeString(keys.Private)
	if err != nil {
		return errors.New("invalid Private key")
	}

	_, err = base64.RawURLEncoding.DecodeString(keys.Public)
	if err != nil {
		return errors.New("invalid Public key")
	}
	return nil
}

//verifyClaims will verify the claims of JWT string
func verifyClaims(claims map[string]interface{}) error {
	//Validate claims
	// sub: The “Subscriber” a mailto link for the administrative contact for this feed.
	// It’s best if this email is not a personal email address,
	// but rather a group email so that if a person leaves an organization,
	// is unavailable for an extended period, or otherwise can’t respond, someone else on the list can.
	if _, ok := claims["sub"]; ok {
		if !(strings.HasPrefix(claims["sub"].(string), "mailto:")) && !(strings.HasPrefix(claims["sub"].(string), "https://")) {
			return errors.New("“Subscriber” claim (sub) is invalid, it should be an email or contact URL")
		}
	}

	//exp : “Expires” this is an integer that is the date and time that this VAPID header should remain valid until.
	// It doesn’t reflect how long your VAPID signature key should be valid, just this specific update.
	// It can be no more than 24 hours
	if _, ok := claims["exp"]; ok {
		now := time.Now().Unix()
		tomorrow := time.Now().Add(24 * time.Hour).Unix()
		if now > claims["exp"].(int64) {
			return errors.New("expiry claim (exp) already expired")
		}
		if claims["exp"].(int64) > tomorrow {
			return errors.New("expiry claim (exp) maximum value is 24 hours")
		}
	}
	return nil
}

func generateJWTSignature(keys VAPIDKeys, JWTInfoAndData string) (string, error) {
	// Signature is the third part of the token, which includes the data above signed with the private key
	// Preparing ecdsa.PrivateKey for signing
	privKeyDecoded, err := base64.RawURLEncoding.DecodeString(keys.Private)
	if err != nil {
		return "", errors.New("invalid VAPID private key string, cannot decode it")
	}

	curve := elliptic.P256()
	px, py := curve.ScalarMult(
		curve.Params().Gx,
		curve.Params().Gy,
		privKeyDecoded,
	)

	pubKey := ecdsa.PublicKey{
		Curve: curve,
		X:     px,
		Y:     py,
	}

	// Private key
	d := &big.Int{}
	d.SetBytes(privKeyDecoded)

	privKey := &ecdsa.PrivateKey{
		PublicKey: pubKey,
		D:         d,
	}

	// Get the key
	hash := crypto.SHA256
	hasher := hash.New()
	hasher.Write([]byte(JWTInfoAndData))

	// Sign JWTInfo and JWTData using the private key
	r, s, err := ecdsa.Sign(rand.Reader, privKey, hasher.Sum(nil))
	if err != nil {
		return "", errors.New("err singing data")
	}

	curveBits := privKey.Curve.Params().BitSize

	if curveBits != 256 {
		return "", errors.New("curveBits should be 256")
	}

	keyBytes := curveBits / 8
	if curveBits%8 > 0 {
		keyBytes++
	}

	rBytes := r.Bytes()
	rBytesPadded := make([]byte, keyBytes)
	copy(rBytesPadded[keyBytes-len(rBytes):], rBytes)

	sBytes := s.Bytes()
	sBytesPadded := make([]byte, keyBytes)
	copy(sBytesPadded[keyBytes-len(sBytes):], sBytes)

	out := append(rBytesPadded, sBytesPadded...)

	return "." + strings.TrimRight(base64.URLEncoding.EncodeToString(out), "="), nil
}

//GenerateVAPIDAuth will generate Authorization header for web push notifications
func GenerateVAPIDAuth(keys VAPIDKeys, claims map[string]interface{}) (string, error) {

	//Validate VAPID Keys
	if err := validateVAPIDKeys(keys); err != nil {
		return "", err
	}

	//Verify Claims
	if err := verifyClaims(claims); err != nil {
		return "", err
	}

	// JWTInfo is base64 Encoded {"typ":"JWT","alg":"ES256"} which is the first part of the JWT Token
	JWTInfo := "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9."

	// JWTData is the second part of the token which contains all the claims encoded in base64
	jsonValue, err := json.Marshal(claims)
	if err != nil {
		return "", errors.New("Marshaling Claims JSON failed" + err.Error())
	}
	JWTData := strings.TrimRight(base64.URLEncoding.EncodeToString(jsonValue), "=")

	JWTSignature, err := generateJWTSignature(keys, JWTInfo+JWTData)
	if err != nil {
		return "", err
	}

	//Compose the JWT Token string
	JWTString := JWTInfo + JWTData + JWTSignature

	// Construct the VAPID header
	VAPIDAuth := fmt.Sprintf(
		"vapid t=%s, k=%s",
		JWTString,
		keys.Public,
	)

	return VAPIDAuth, nil
}
